### PR TITLE
Add GitHub workflow for releases

### PR DESCRIPTION
**Pull request by mitchnielsen**
## Summary

Adds a new GitHub workflow for releases that will:
- Upload the CRDs and default CR manifest
- Build and push the docker image

Closes https://linear.app/prefect/issue/PLA-220/design-release-process

Closes https://github.com/PrefectHQ/prefect-operator/issues/18

## Testing

- `docker pull prefecthq/prefect-operator-dev:pr-58`

### CODE REVIEW

**Comment by mitchnielsen** [comment]
Still pretty new to GitHub Actions so going to try to simulate a release here to ensure this all works. Welcoming any suggested changes here too.

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      contents: write
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: jdx/mise-action@v2
+      - name: Build and upload manifests
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          cd config/manager && kustomize edit set image controller=${{ github.ref_name }}
```
Comment (mitchnielsen) [inline .github/workflows/release.yaml:22]: We'll need to update this and/or the manifest's `image` field to match the image registry value.

File: .github/workflows/release.yaml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    steps:
+      - uses: actions/checkout@v4
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+
+      - name: Login to Docker Hub
+        uses: docker/login-action@v3
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
```
Comment (mitchnielsen) [inline .github/workflows/release.yaml:38]: Will need to populate these.

File: .github/workflows/release.yaml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+      - name: Build and push
+        uses: docker/build-push-action@v6
+        with:
+          context: .
+          push: true
+          tags: prefect/prefect-operator:${{ github.ref_name }}
```
Comment (mitchnielsen) [inline .github/workflows/release.yaml:45]:
```suggestion
          tags: <registry.domain.here>/prefect/prefect-operator:${{ github.ref_name }}
```

File: .github/workflows/release.yaml (hunk 1 above)
Comment (chrisguidry) [reply .github/workflows/release.yaml:22]: Yep, we can do `prefecthq/prefect-operator` (https://hub.docker.com/u/prefecthq)

Comment (mitchnielsen) [reply .github/workflows/release.yaml:22]: Sounds good, will set this up after the long weekend 👍🏼

File: .github/workflows/release.yaml (hunk 2 above)
Comment (parkedwards) [reply .github/workflows/release.yaml:38]: @mitchnielsen we can add them in Terraform.  i'll DM you

Comment (mitchnielsen) [reply .github/workflows/release.yaml:38]: https://github.com/PrefectHQ/platform/pull/7409

**Comment by mitchnielsen** [comment]
The pull request scenario looks good, and after merging we can test a real release workflow with something like `v0.0.1` to ensure that all works as well.

File: .github/workflows/release.yaml [hunk 4]
```diff
@@ -0,0 +1,71 @@
+# Prefect Operator release workflow.
+name: Release
+
+on:
```
Comment (jamiezieziula) [inline .github/workflows/release.yaml:4]:
```suggestion
"on":
```

File: .github/workflows/release.yaml [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  release:
+    types: [published]
```
Comment (jamiezieziula) [inline .github/workflows/release.yaml:6]: i'm fine either way, but i think we should be consistent with our lists. either brackets or with dashes as done below

**Review by chrisguidry — APPROVED** [review]
Looks great!  I really appreciate the `-dev` image support and the release note setup too 🙇‍♂️

File: .github/workflows/release.yaml [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+  build-and-push-docker-image:
+    runs-on: ubuntu-latest
+    # The GitHub environments are created by Terraform and map to Docker Hub repositories:
+    # - dev: https://hub.docker.com/r/prefecthq/prefect-operator-dev
+    # - prod: https://hub.docker.com/r/prefecthq/prefect-operator
+    # The environment will be 'prod' if the GitHub event is a release. Otherwise, it will be 'dev'.
+    environment: ${{ github.event_name == 'release' && 'prod' || 'dev' }}
+    steps:
+      - uses: actions/checkout@v4
```
Comment (jamiezieziula) [inline .github/workflows/release.yaml:40]: maybe a name here? idk if necessary tho

**Review by jamiezieziula — APPROVED** [review]
a few nits but otherwise looks good

File: .github/workflows/release.yaml (hunk 5 above)
Comment (chrisguidry) [reply .github/workflows/release.yaml:6]: Maybe we should throw the `yamllint` pre-commit on this repo?  There's a ton of YAML in here

File: .github/workflows/release.yaml [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        uses: docker/metadata-action@v5
+        with:
+          images: prefecthq/prefect-operator${{ github.event_name == 'pull_request' && '-dev' }}
+          tags: |
+            type=ref,event=pr
+            type=semver,pattern={{version}},event=tag
+          labels: |
+            org.opencontainers.image.title=prefect-operator
+            org.opencontainers.image.description=Prefect Operator image
+            org.opencontainers.image.vendor=PrefectHQ
```
[the remark this answers is not in the document]
Comment (mitchnielsen) [reply .github/workflows/release.yaml:63]: Ah yeah! I'm up for anything. I grabbed this from the [Prefect Prometheus exporter workflow](https://github.com/PrefectHQ/prometheus-prefect-exporter/blob/4fe99739ac77ddd356c65cb51bbf0cce07ab8cfa/.github/workflows/release-build-deploy.yaml#L44).

Looking at [the spec](https://specs.opencontainers.org/image-spec/annotations/#pre-defined-annotation-keys), the company name goes here so the `HQ` suffix may not make sense.

[the remark this answers is not in the document]
Comment (chrisguidry) [reply .github/workflows/release.yaml:63]: Yeah, let's drop "HQ" from both, then, right?  Not critical right now for sure

File: .github/workflows/release.yaml (hunk 4 above)
Comment (mitchnielsen) [reply .github/workflows/release.yaml:4]: Down for this, any reason why though? The [docs](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#on) don't seem to do it this way but maybe YAML keywords bite us sometimes?

File: .github/workflows/release.yaml [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      - name: Docker image metadata
+        id: metadata
+        uses: docker/metadata-action@v5
+        with:
+          images: prefecthq/prefect-operator${{ github.event_name == 'pull_request' && '-dev' }}
+          tags: |
+            type=ref,event=pr
+            type=semver,pattern={{version}},event=tag
```
Comment (parkedwards) [inline .github/workflows/release.yaml:59]: where is this {{version}} template var coming from?

Comment (parkedwards) [inline .github/workflows/release.yaml:59]: are these tags all being attached on every build?

**Review by parkedwards — APPROVED** [review]
LGTM

File: .github/workflows/release.yaml (hunk 8 above)
Comment (mitchnielsen) [reply .github/workflows/release.yaml:59]: It's apparently built in magic from https://github.com/docker/metadata-action?tab=readme-ov-file#semver. We use this in the [Prefect Prometheus release setup](https://github.com/PrefectHQ/prometheus-prefect-exporter/blob/4fe99739ac77ddd356c65cb51bbf0cce07ab8cfa/.github/workflows/release-build-deploy.yaml#L40) as well.

File: .github/workflows/release.yaml (hunk 5 above)
Comment (mitchnielsen) [reply .github/workflows/release.yaml]: Good shout, updated in 519e4af

File: .github/workflows/release.yaml (hunk 6 above)
Comment (mitchnielsen) [reply .github/workflows/release.yaml]: Yeah always helps, updated in 9fb1bf0

File: .github/workflows/release.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (mitchnielsen) [reply .github/workflows/release.yaml]: Updated in ebc08b4

**Comment by chrisguidry** [comment]
Oh one quick thing, can we make the release jobs depend on the tests?

**Comment by mitchnielsen** [comment]
> Oh one quick thing, can we make the release jobs depend on the tests?

Good shout, looking into this now

File: .github/workflows/release.yaml (hunk 4 above)
Comment (jamiezieziula) [reply .github/workflows/release.yaml:4]: i think yamllint with throw an error. 'on' is a truthy value in yaml iirc. it won't break anything either way

Comment (mitchnielsen) [reply .github/workflows/release.yaml:4]: oh interesting, yeah I found https://github.com/adrienverge/yamllint/issues/430. weird one - will change

File: .github/workflows/release.yaml (hunk 5 above)
Comment (mitchnielsen) [reply .github/workflows/release.yaml]: Added linting setup in 0ccd597875298df.

File: .github/workflows/release.yaml (hunk 4 above)
Comment (mitchnielsen) [reply .github/workflows/release.yaml]: Updated in 0ccd597875298df.

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+      - id: trailing-whitespace
+
+  - repo: https://github.com/adrienverge/yamllint.git
+    rev: v1.28.0
+    hooks:
+      - id: yamllint
+        args:
+          - --strict
+
+  - repo: https://github.com/rhysd/actionlint
```
Comment (chrisguidry) [inline .pre-commit-config.yaml:23]: Nice find!

**Review by chrisguidry — APPROVED** [review]

**Pull request merged**